public/src/components/channelManagement/richTextEditor.tsx

```diff
@@ -0,0 +1,407 @@
+            // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+            onClick={() => {
+              chain
+                .toggleItalic()
+                .focus()
+                .run();
+            }}
+          >
+            Italic
+          </button>
+          <span className="remirror-button-spacer">&nbsp;</span>
+          <button
+            className="remirror-button"
+            // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+            onClick={() => {
+              chain.insertText('%%ARTICLE_COUNT%%').run();
+            }}
+          >
+            Articles
+          </button>
+          <button
+            className="remirror-button"
+            // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+            onClick={() => {
+              chain.insertText('%%CURRENCY_SYMBOL%%').run();
+            }}
+          >
+            Currency
+          </button>
+          <button
+            className="remirror-button"
+            // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+            onClick={() => {
+              chain.insertText('%%COUNTRY_NAME%%').run();
+            }}
+          >
+            Country
+          </button>
+        </>
+      )}
+    </div>
+  );
+};
+
+// Helper function - converts an array of strings into a set of (stringified) HTML <p> elements
+const parseCopyForParagraphs = (copy: string[]): string => {
+  let res = '';
+
+  copy.forEach(paragraph => {
+    res += `<p>${paragraph}</p>`;
+  });
+  return res;
+};
+
+export const getRteCopyLength = (copy: string[]): number => {
+  let paragraphsCheck = copy.join('');
+
+  paragraphsCheck = paragraphsCheck.replace(/<.*?>/g, '');
+  paragraphsCheck = paragraphsCheck.replace(/%%CURRENCY_SYMBOL%%/g, ' ');
+  paragraphsCheck = paragraphsCheck.replace(/%%ARTICLE_COUNT%%/g, '     ');
+  paragraphsCheck = paragraphsCheck.replace(/%%COUNTRY_NAME%%/g, '          ');
+
+  return paragraphsCheck.length;
+};
+
+// Component function
+const RichTextEditor: React.FC<RichTextEditorProps> = ({
+  classes,
+  disabled,
+  label,
+  helperText,
+  name,
+  error,
+  updateCopy,
+  copyData,
+}: RichTextEditorProps) => {
+  // Make sure the supplied copy is in an Array, for processing
+  if (copyData == null) {
+    copyData = [];
+  }
+  if (!Array.isArray(copyData)) {
+    copyData = [copyData];
+  }
+
+  // Instantiate the Remirror RTE component
+  const { manager, state } = useRemirror({
+    extensions: () => [
+      new BoldExtension(),
+      new ItalicExtension(),
+      new LinkExtension({ autoLink: true }),
+      new TextHighlightExtension(),
+    ],
+    content: parseCopyForParagraphs(copyData),
+    selection: 'start',
+    stringHandler: 'html',
+  });
+
+  // Handle to the ProseMirror <div> element containing all the RTE text
+  let ref: HTMLDivElement | null = document.createElement('div');
```

Comment:
This feels like it goes against the spirit of remirror. I need to read the docs more but I think we should use their api to listen for events and get the content.
The docs are a bit light but it seems you can use the `EventsExtension` to listen for the blur event, e.g:
```
const eventsExtension = new EventsExtension();
eventsExtension.addHandler('blur', () => { })
// add eventsExtension to extensions array
```

And I think you can use getHTML -
https://remirror.io/docs/getting-started/commands-and-helpers/#helpers

Reply:
The current extraction code is the code I'm least happy with. Will look at the EventsExtension as suggested.

Reply:
In terms of getting the content out, here's the example hook which you pass into the Remirror component:
```
const hooks = [
  () => {
    const { getHTML } = useHelpers();

    const handleSaveShortcut = useCallback(
      ({ state }) => {
        console.log(JSON.stringify(getHTML(state)));
        return true; // Prevents any further key handlers from being run.
      },
      [getText],
    );

    // "Mod" means platform agnostic modifier key - i.e. Ctrl on Windows, or Cmd on MacOS
    useKeymap('Mod-s', handleSaveShortcut);
  },
];
```
The challenge is to trigger this onBlur rather than on key press, and I'm not sure how to do that with the `EventsExtension`

Reply:
Actually, I've found a way to use the EventsExtension in the hook-
```
const hooks = [
  () => {
    const { getText, getHTML } = useHelpers();

    const handleSaveShortcut = useCallback(
      ({ state }) => {
        console.log(JSON.stringify(getHTML(state)));
        return true; // Prevents any further key handlers from being run.
      },
      [getHTML],
    );
    manager.getExtension(EventsExtension).addHandler('blur', handleSaveShortcut);
  },
];
```

Reply:
Fix applied. I've also had to go through the code base adding in the `paragraphs[]` attribute alongside the deprecated `messageText` attribute - which will need to be removed from the code base after we update all the LIVE/CODE/DEV banner test JSON files to use`paragraphs[]`